[Security] Added docs about the authenticator-based Security system

wouterj opened the pull request:
As the new system is experimental, I think it's best to not update the main documentation of Security yet (also, the main guide isn't really affected by the new system - except from the removal of anonymous). This can be done once we deprecate the old system (in 5.3 or 5.4).

I do however think we should link more to this article, as I want to push people to use it. What do you think about adding a `.. tip::` in:

* [x] `security/guard_authentication` & `security/form_login_setup`
* [ ] <s>`security/custom_authentication_provider` (replace the one to Guard)</s>
* [x] `security` (at the start, as a "hey, please start by doing `enable_authenticator_manager: true`")

Fixes #13575

wouterj committed:
[Security] Added docs about the authenticator-based Security system

wouterj commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+    use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+    use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+
+    class ApiKeyAuthenticator extends AbstractAuthenticator
```
I'm not sure I like having an API token authenticator, as it doesn't use the default `Passport` and has no success response. Doing a login form isn't great either (users should extend `AbstractFormLoginAuthenticator` then).
Does anyone have an idea for a good example authenticator that we can use?

noniagriconomie replied:
What about a magic link auth?
Retrieves and checks data from request (ie a user uniq token)
If ok, a response that redirects to a default authenticated page

wouterj replied:
I really liked this idea, until I realized this also used the `SelfValidatingPassport`. In the end, I think we can keep the current example for now. Let's see if it's confusing.

wouterj commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+            // ...
+        ]);
+
+The new system is backwards compatible with the current authentication
```
We should maybe come up with a consistent name for the new system. "Guards" were very nice to use in a sentence. "Authenticator-based system" is both too technical and too long imho. "New system" on the other hand is too vague.

noniagriconomie replied:
If you use the word « new » maybe use the word « old » instead of « current » ?
Otherwise, what about « next » instead of new?
As i understood, it will be the new when this will be the only way, isnt it?

wouterj replied:
It's the new system, but it's experimental so it's not really the "current" system. Decided to keep using "new" or "authenticator-based system" - at least for now.

noniagriconomie commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+
+In Symfony 5.1, a new authentication system was introduced. This system
+changes the internals of the Security system, to make it more extensible
+and easier to understand.
```
What about: more understandable
As imo, the code for an external pov can be not this easy :)

noniagriconomie commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+
+Authenticators should implement the
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.
+The easiest way to use it is by extending
```
What about « An alternative could be to extends the abstract ... »

Because an abstract is not always easier imo, sometimes implementing fully is easier :)

noniagriconomie commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+    :doc:`/security/remember_me` for more information.
+
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\Passport\\Badge\\PasswordUpgradeBadge`
+    This badge requires the plaintext password and a password upgrader.
```
I would write first the aim, then the requirement like you did for others

jmsche commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,432 @@
+            // ... get the $user from the $username and validate no
+            // parameter is empty
+
+            return new Password($user, new PasswordCredentials($password), [
```
I guess it should be `new Passport`?

javiereguiluz commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+Using the new Authenticator-based System
```
"system" looks too generic to me ... what about `Using the new Authenticator-based Security` ?

javiereguiluz commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+login and an API token authentication). In these cases, it is now required
+to configure the *authentication entry point*. The entry point is used to
+generate a response when the user is not yet authenticated. This can be
+used to for instance redirect the user to the login page.
```
`used to for instance redirect` -> `used for instance to redirect` ?

javiereguiluz commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+generate a response when the user is not yet authenticated. This can be
+used to for instance redirect the user to the login page.
+
+You can configure this using the ``entry_point`` config setting:
```
`config setting` -> `option` ?

javiereguiluz commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+    }
+
+The ``authenticate()`` method is the most important method of the
+authenticator. It's job is to extract credentials (e.g. username &
```
`It's job` -> `Its job`

javiereguiluz reviewed (approved):
Wouter, thanks for contributing these docs. They are easy to follow even if you don't know much about security.

pskt commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+
+Authenticators should implement the
+:class:`Symfony\\Component\\Security\\Http\\Authenticator\\AuthenticatorInterface`.
+You can also
```
```suggestion
You can also extend
```

wouterj commented:
Thanks for the review. I've fixed the comments and also added some references to the article from other places in the docs (as proposed in the description).

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+
+.. versionadded:: 5.1
+
+    Authenticator-based security was introduced as
```
```suggestion
    Authenticator-based security was introduced as an
```

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+                ['path' => '^/admin/login', 'roles' => AccessListener::PUBLIC_ACCESS],
+
+                // but require authentication for all other admin routes
+                ['path' => '^/admin', 'roles' => AccessListener::PUBLIC_ACCESS],
```
The role is wrong here

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+Sometimes, one firewall has multiple ways to authenticate (e.g. both a form
+login and an API token authentication). In these cases, it is now required
+to configure the *authentication entry point*. The entry point is used to
+generate a response when the user is not yet authenticated. This can be
```
```suggestion
generate a response when the user is not yet authenticated but tries to access a page that
requires authentication. This can be
```

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+    You can also create your own authentication entry point by creating a
+    class that implements
+    :class:`Symfony\\Component\\Security\\Http\\EntryPoint\\AuthenticationEntryPointInterface`.
+    You can then set ``entry_point`` to the FQCN (e.g.
```
Hmm, I'm wondering if FCQN should say "service id" to be more accurate. But we would still show a class name in the example below.

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+    to also update these when you're refactoring your application to the
+    new system.
+
+    The new authenticator interface has quite some overlap with the guard
```
```suggestion
    The new authenticator interface has many similarities with the guard
```

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+
+    .. code-block:: terminal
+
+        $ composer remove symfony/security-guard
```
Will that work if you only have, for example, `symfony/security-bundle` in your `composer.json`? That is the most likely use case, as `composer require security` gives you that package.

wouterj replied:
Oh yes sorry, really good catch. This should be removed, as sadly `symfony/security-bundle` depends on the guard package.

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+         * to be skipped.
+         *
+         * Returning `null` indicates this authenticator supports to be called
+         * lazily.
```
Even for me, I'm not sure what to do with this information :p. Is it relevant for users? If so, what does it *really* mean? When should I return null?

wouterj replied:
It can be very relevant, but let's remove it and discuss it in a cache/lazy firewall related section instead (if we ever write one).

weaverryan commented on security/experimental_authenticators.rst:
```diff
@@ -0,0 +1,434 @@
+
+Passports are introduced by the authenticator system. A passport contains
+all information that should be used for authentication. This object is used
+to check authentication by the security system.
```
Hmm, these 3 sentences don't really *say* much :). Maybe just:

> A passport is an object that contains the user that will be authenticated as well
> as other pieces of information, like whether a password should be checked or
> if "remember me" functionality should be enabled.

weaverryan reviewed (commented):
Minor stuff - pretty sweet :)
